README.md

```diff
@@ -114,17 +115,23 @@ const resolvers = {
     publishArticle: (object, args, context, info) => {
       const user = context.kauth.accessToken.content // get the user details from the access token
       return Database.createArticle(args.title, args.content, user)
+    },
+	unpublishArticle: (object, args, context, info) => {
+	  const user = context.kauth.accessToken.content
+      return Database.deleteArticle(args.title, user)
     }
   }
 }
 
+const keycloak = new Keycloak()
+
 const server = new ApolloServer({
   typeDefs: [KeycloakTypeDefs, typeDefs], // 1. Add the Keycloak Type Defs
   schemaDirectives: KeycloakSchemaDirectives, // 2. Add the KeycloakSchemaDirectives
   resolvers,
   context: ({ req }) => {
     return {
-      kauth: new KeycloakContext({ req }) // 3. add the KeycloakContext to `kauth`
+      kauth: new KeycloakContext({ req }, keycloak) // 3. add the KeycloakContext to `kauth`
```

Comment:
Any reason for this change?

Reply:
An object keycloak is added into KeycloakContext constructor in order to support hasPermission functionality. Anyhow, it's an optional parameter and if it's not supplied hasPermission always denies access.  An example how to use hasPermission is given in this code fragment and that is the reason why keycloak object is passed into KeycloakContext construtor.

Reply:
There are two factors here:

- backwards compatibility
- side effects (very important in security libs)

We are good in backwards compatibility business but not sure if we want to get into (undocumented), feature is disabled if you do not pass x or y. Especiallty that this is keycloak. 

If you feel that it is user responsibility to create keycloak instance then let's make hard change and require this. I feel that we might simply change examples to not pass keycloak. Then if it is not passed then context can create it. I'm not sure if we want to have some side effects that are not obvious - some feature disabled if keycloak is not passed etc

I would hate some complexity on public API that really generates matrix of test cases we need to cover.

Reply:
As author you had the vision so would love to see what would be your take on this.

Reply:
Good point about side effects.

My vision was to provide solution for hasPermission without any breaking changes in the library. The outcome is a side effect with hasPermission function that always denies access, if the keycloak parameter is not supplied. This is not desirable behavior in security libraries, I can agree with you.

The problem is that if we want to implement hasPermission function, we need Keycloak Authorization Services REST API, which is available via Keycloak instance. Because Keycloak instance is created using Keycloak configuration, which is not available within KeycloakContext, we can't create it, so it  has to be supplied as a parameter. 

Knowing this I would prefer to go with required parameter, because it's not possible to created a Keycloak instance within context. This however means breaking changes. Additionally, users who only use role-based access control will need to provide keycloak object as well, despite they don't need this object.

I was thinking about alternative solution, but we will end up with the same issue of obscure context behavior when permissions data are not accessible.

Reply:
> Knowing this I would prefer to go with required parameter, because it's not possible to create a Keycloak instance within context. This however means breaking changes. Additionally, users who only use role-based access control will need to provide keycloak object as well, despite they don't need this object.


I think this is good. Library is still in 0.x so we can do breaking changes (and document them properly). Secondly - library was using keycloak anyway but relying on global object added to request. That was due to some keycloak limitations that have been fixed (we made PR's). 

I do think that after explanation the best decision was made and we should merge it as it is. 
We can refactor it later for role-based stuff to use keycloak as well.